### PR TITLE
Option to read hosts from terraform output json

### PR DESCRIPTION
**Pull request by jakolehm**
Improves integration with Terraform.

### main.tf:
```tf
variable "auth_token" {
  description = "Packet authentication token"
}

variable "project_id" {
  description = "Packet project id"
}

provider "packet" {
  auth_token = "${var.auth_token}"
}

resource "packet_device" "pharos_master" {
  hostname         = "pharos-master"
  plan             = "baremetal_0"
  facility         = "ams1"
  operating_system = "ubuntu_16_04"
  billing_cycle    = "hourly"
  project_id       = "${var.project_id}"
}

resource "packet_device" "pharos_worker" {
  count            = 3
  hostname         = "pharos-worker-${count.index}"
  plan             = "baremetal_0"
  facility         = "ams1"
  operating_system = "ubuntu_16_04"
  billing_cycle    = "hourly"
  project_id       = "${var.project_id}"
}

output "pharos" {
  value = {
    masters = {
      address         = "${packet_device.pharos_master.*.network.0.address}"
      private_address = "${packet_device.pharos_master.*.network.2.address}"
      role            = "master"
      user            = "root"
    }

    workers = {
      address         = "${packet_device.pharos_worker.*.network.0.address}"
      private_address = "${packet_device.pharos_worker.*.network.2.address}"

      label = {
        foo = "bar"
      }

      user = "root"
      role = "worker"
    }
  }
}
```

### cluster-without-hosts.yml

```yaml
network:
  pod_network_cidr: 172.32.0.0/16
  trusted_subnets:
    - 10.80.0.0/16
addons: {}
```

### Usage 
```sh
$ terraform apply
$ terraform output -json > tf.json
$ pharos-cluster up -c cluster-without-hosts.yml --hosts-from-tf ./tf.json
```

### CODE REVIEW

**Comment by jakolehm** [comment]
@kke imo it would make sense to pipe output from `terraform output` directly to `pharos-cluster` but stdin read assumes currently full cluster.yml. Any ideas how to improve?

**Comment by matti** [comment]
This looks great!  The only thing that concerns me is that this assumes that all workers have the same labels and connection settings. How about setting the workers as a list of those maps:

```
    workers = [
      {
          address         = "${packet_device.pharos_hard_worker.*.network.0.address}"
         ...
      },
     {
          address         = "${packet_device.pharos_slow_worker.*.network.0.address}"
          ...
      }
    ]
  }
```

?

**Comment by jakolehm** [comment]
Maybe my example was a bit too simplified.. but you can already have different "groups" of workers.

```tf
output "pharos" {
  value = {
    masters = {
      address         = "${packet_device.pharos_master.*.network.0.address}"
      private_address = "${packet_device.pharos_master.*.network.2.address}"
      role            = "master"
      user            = "root"
    }

    type_0 = {
      address         = "${packet_device.pharos_worker.*.network.0.address}"
      private_address = "${packet_device.pharos_worker.*.network.2.address}"

      label = {
        metal = "type0"
      }

      user = "root"
      role = "worker"
    }

    type_2 = {
      address         = "${packet_device.pharos_worker2.*.network.0.address}"
      private_address = "${packet_device.pharos_worker2.*.network.2.address}"

      label = {
        metal = "type2"
      }

      user = "root"
      role = "worker"
    }
  }
}
```

**Comment by matti** [comment]
ok! didnt realize that is was this fancy

**Comment by kke** [comment]
Defaulting to `./cluster.yml` is a bit problematic here. It's difficult to figure out if it should it read it first and merge the stuff from stdin or should input from stdin make it skip trying to read the default file at all? 

Also maybe the config path could be a param instead of an option.

If the config param would be multivalued and it would `config_list.map {|x| parse(x)}.reduce(:merge)`, then you could do something like `terraform --foo | pharos-cluster up -c cluster.yml -c network_settings.yml`. Of course this would require that it can detect which inputs to pass through the terraform parser (which i guess could be done by just checking if the first char is `{`).

**Comment by jakolehm** [comment]
> Defaulting to ./cluster.yml is a bit problematic here.

I agree, should we drop the default?

**Comment by kke** [comment]
Depends a bit how it's supposed to be used.

If the idea is to have something like:
```
addons/foofoo/foo.rb
cluster.yml
```

in git repos, then maybe it makes sense.

**Comment by kke** [comment]
Or maybe default to `./cluster.yml` only as the final last solution if there's no param and nothing in the stdin.

But this too will be confusing if you're used to do `pharos-cluster up` and then want to pipe the terraform stuff on top, you need to specify the `cluster.yml` as option explicitly.

**Comment by jakolehm** [comment]
Ok, maybe stdin change should not be part of this PR. IMO reading from json file is good enough for most use cases.

File: lib/pharos/up_command.rb [hunk 1]
```diff
@@ -22,13 +31,47 @@ def default_config_yaml
 
     def execute
       puts pastel.green("==> Reading instructions ...")
-      configure(load_config)
+      config_hash = load_config
+      if hosts_from_tf
+        puts pastel.green("==> Importing hosts from Terraform ...")
+        config_hash['hosts'] = load_tf_json
```
Comment (SpComb) [inline lib/pharos/up_command.rb]: Perhaps a merge rather than a replace?

Comment (jakolehm) [reply lib/pharos/up_command.rb]: Now it appends to hosts array.

**Pull request merged**